src/lin.py

```diff
@@ -1,16 +1,60 @@
 """Implementation of a linear time exact matching algorithm."""
 
 import argparse
+from asyncio.base_subprocess import ReadSubprocessPipeProto
+import numpy as np
 
+from utils import *
+
+
+
+def borderArray(p, x):
+    # Step 1: Construct s = p$x and its borderarray b
+    s = p + "$" + x
+    ba = np.zeros(len(s), dtype=int)
+    for i in range(1, len(s)):
+        b = ba[i - 1]
+        while b > 0 and s[i] != s[b]:
+            b = ba[b - 1]
+        if s[i] == s[b]:
+            ba[i] = b + 1
+        else:
+            ba[i] = 0
+    return ba, s
+
+
+def occurrence(p, x):
+    # Step 2: Report an occurence of p in x at position i-2m if b[i]=m
+    idx = []
+    ba = borderArray(p, x)[0]
+    s = borderArray(p, x)[1]
+    for i in range(len(ba)):
+        if ba[i] == len(p):
+            idx.append(i - 2 * len(p))
+    return idx
+
+def lin(fasta, fastq):
+    #Linear border array that outputs simple SAM format
+    ps, ps_names = parse_fastq(fastq)
+    xs, xs_names = fasta_recs(fasta)
+
+    for i in range(len(ps)):
+        for j in range(len(xs)):
+            idx = occurrence(ps[i], xs[j])
+            for k in idx:
+                e = get_edits(ps[i], xs[j][k :(k + len(ps[i])-1)])[2]
+                l = edits_to_cigar(e)
```

Comment:
We are doing exact pattern matching, so the CIGAR will always be `f"{len(p)}M"`, so this bit isn't necessary in this function.